### ISSUE TITLE
Streaming batch normalization?

### ISSUE THREAD

**Issue by tu1258** [google-research/google-research#356]
This question is related to https://github.com/google-research/google-research/tree/master/kws_streaming and https://arxiv.org/pdf/2005.06720.pdf
Batch normalization calculate average and variance and apply it to entire feature map. So the entire feature map is changed, the convolution filter has to compute the entire feature map. Then how is that "streamable" and how do you do it? Thanks!

**Comment by rybakov** [comment]
Batch normalization is applied on feature domain, so it is streamable by default. We apply streaming conversion only on layers, which do data processing in time domain (such as conv, rnn,...). For example during conversion of conv layer to streaming mode we insert time buffer before conv layer, so it can run convolution incrementally on new input samples. All other layers which do processing in feature domain only are kept as it is.

If you do not need to do folding of conv layer with batch normalization then it should work out of the box. But if you need to fold conv layer with batch norm, then there will be several steps: 1) convert non streaming model to streaming one, 2) fold conv and batch norm layers of streaming model. For second step I would suggest to use keras tool: [model_transformer](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/quantization/keras/graph_transformations/model_transformer_test.py). Please let me know if you have more questions, feel free to create an example, so it will be clear what kind of problem you have with kws_streaming lib.

**Comment by tu1258** [comment]
Hi, @rybakov  Thanks for your reply! So what you are saying is batch normalization only affects the latest column of feature map. (the gray area in the [paper](https://arxiv.org/pdf/2005.06720.pdf)) Bacth normalization is applied on feature domain so it updates the feature map column by column. Please correct me if I misunderstand. Much appreciated!

**Comment by rybakov** [comment]
You are saying that: "_However, batch normalization is computed by moving average when inference thus I am confused now._"
It looks like you a mixing batch normalization behavior during training and inference. 

Yes, batch normalization computes moving average of the input features, but only during training.
But in the inference mode all weights of batch normalization layer are frozen. So, during inference batch normalization can be represented as a simple 1x1 convolution as described in [link](https://tehnokv.com/posts/fusing-batchnorm-and-conv/), 1x1 convolution in this case is just elementwise multiplication with addition of bias and it is streamable by default, also can be folded/fused with predecessing conv layer.

Please find attached example colab with conv and batch norm streaming (it is an implementation of the model on your drawing): 
[conv_batch_norm.zip](https://github.com/google-research/google-research/files/5077058/conv_batch_norm.zip)
. It demonstrates how to train a model a non streaming model, run inference in non streaming mode with TF/TFLite and run inference in streaming mode with TF/TFLite (data sets in this example are artificially generated and used only to validate model correctness).
In this colab you will see that during training conv layer is producing output with dim: [16/_batch_, 47/_time_, 18/_feature1_, 5/_feature2_] and batch norm layer is processing  [16/_batch_, 47/_time_, 18/_feature1_, 5/_feature2_] (it applies exponential averaging only during training).
During streaming inference _batch_ and and _time_ dims are equal to one, so conv layer produces output [1/_batch_, 1/_time_, 18/_feature1_, 5/_feature2_] and batch norm layer is processing  [1/_batch_, 1/_time_, 18/_feature1_, 5/_feature2_]. The weights of batch norm layer are frozen during inference and applied only on feature dimensions (so there is no exponential averaging during inference).

Thank you for looking into kws_streaming lib! and let me know if you have more questions.
/Oleg.

**Comment by rybakov** [comment]
Added couple of examples:

Example of folding/fusing convolution layer with batch normalization [conv_batch_norm_folding](https://github.com/google-research/google-research/blob/master/kws_streaming/colab/conv_batch_norm_folding.ipynb)

Here is end to end test of running convolution with batch normalization with TF and TFLite [conv_batch_norm_on_generated_data](https://github.com/google-research/google-research/blob/master/kws_streaming/colab/conv_batch_norm_on_generated_data.ipynb)

**Comment by tu1258** [comment]
Hi, @rybakov 
I look at the code of batch norm but it seems that batch norm is calculated by moving average in inference mode? https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/layers/normalization.py#L763-L791
Thanks!

**Comment by rybakov** [comment]
Summary: there is no any streaming during training: we train a model as it is. 
Batch normalization during training compuets moving average of features in time and is updated at every mini batch.

During inference of batch normalization layer, we use moving average value, which was pre-computedd during training (but now it is frozen - it is not updated any more, as shown in keras [normalization.py](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/layers/normalization.py#L765)). It is applied on features and there is no averaging in time becasue we are in inference mode now. So this layer is streamable in inference mode by default.
